### ISSUE TITLE
Research of LiteDB concurrency issues

### ISSUE THREAD

**Issue by podkolzzzin** [mbdavid/LiteDB#639]
Hello!
I have found a reason of most of concurrency bugs that I have reported.
I'm sorry that I can't propose you a pull request or solution of this problem, but at least I can explain what is happening, and hope it helps you to resolve this issue.

If cut down all research to one sentence:
There is a possibility when one thread/process can delete file during constructing of FileStream. In this case UnauthorizedAccessException will be thrown.

You expect that if there is no file(I am talking about -journal file) .net FileStream will throw FileNotFoundException and then nobody could delete file until FileStream will be closed/disposed. But unfortunately there is a chance when the file can be deleting during executing of FileStream.ctor.
I have proved it writing this code:
```
    static void Main(string[] args) {

      using (File.Create("f"))
        ;


      for (int i = 0; i < 5; i++)
        new Thread(o).Start();

      for (int i = 0; i < 5; i++)
        new Thread(a).Start();
    }

    private static void a() {

      while (true) {
        try {
          File.Delete("f");
        }
        catch { }
      }
    }

    private static void o() {

      while (true) {
        try {
          using (var fs = new FileStream("f", FileMode.Open, FileAccess.ReadWrite, FileShare.ReadWrite))
            ;
        }
        catch (IOException) { }
        catch (UnauthorizedAccessException ex) {
          Console.WriteLine(ex);
        }
      }
    }
```
 
It means that you have at least 2 race-conditions.
First one:
```
    Thread#1: TransactionService.Recovery is called, during constructing LiteDatabase object
              >>FileDiskService.ReadJournal is called
              >>>>FileDiskService.CreateFileStream is called
              >>>>>>FileStream.ctor is called

    Thread#2: Do all the same that Thread#1
              FileDiskService.ClearJournal is called
              >>FileDiskService.Dispose is called
              >>>>FileStream.Dispose is called
              >>>>FileHelper.TryDelete is called and finished succesfully.

    Thread#1: >>>>>>FileStream.ctor throws UnauthorizedAccessException
```

Second one:
	Quite similar issue during WriteJournal

I hope this information will be helpful for you to fix concurrency issues. I am sorry that can't suggest pull request to fix it.

**Comment by mbdavid** [comment]
Hi @podkolzzzin, thanks man! This make a sense. I will do some testing tomorrow about read/write journal because there are some bugs in this.

**Comment by mbdavid** [comment]
Hi @podkolzzzin, after this i decide to complete remove external journal file. Now, journal will be write direct in same datafile, after used area. Here my logic:

# Commit

1. Lock datafile in exclusive mode (no others can read/write)
2. Get next page from Header.LastPageID (out of used area on disk)
3. Set disk length ("real data" + "journal data")
4. Mark header page as "Recovery = true" (new attribute)
5. Write all journal pages in this non-used area disk (after LastPageID)
6. Starts writing dirty pages in disk starting from Header (page 0)
7. Mark header page as "Recovery = false" and persist in disk
8. Shrink datafile excluding journal space
9. Flush here?
10. Unlock exclusive mode datafile

# Recovery

1. Open datafile (can be opened only if is not in exclusive mode)
2. If Header.Recovery = true must be recovery data. If not, go away
3. Lock datafile in exclusive mode
4. ForEach pages after Header.LastPageID, reading from journal area to physical area
5. Re-Write header page but now marks Recovery = false
6. Flush here?
7. Unlock exclusive mode datafile

I'm working in this on "nojournal" branch if you want take a look.

**Comment by podkolzzzin** [comment]
So, it would be a concept of single-write/multiple-read?

In recovery algorithm you also can use double-check-lock pattern to avoid some extra work/races.
It could be implemented like this:
2. If Header.Recovery = true must be recovery data. If not, go away
3. Lock datafile in exclusive mode
4. If Header.Recovery = true must be recovery data. If not, go away

**Comment by mbdavid** [comment]
LiteDB always works with single-write/multi-read. But current version has a "Reserved" lock that works as "UpgradableRead" state. Removing this state it's quite simple, becase any "BeginTrans()" will lock datafile and will release only after write on disk,
In my first tests works as expected. Now I will try get more examples and unit tests. I will apply this change in major version because I want remove/change features like:

- Remove transaction
- Remove auto-create index on query (work better on read-only database)
- Implement full-scan query execution
- Cache results on query

**Comment by podkolzzzin** [comment]
Sounds really cool! 
Good luck! I shared some code samples in my previous requests that detect some concurrency issues, hope they would be useful.

**Comment by lbnascimento** [comment]
Hi! With the objective of organizing our issues, we are closing old unsolved issues. Please check the latest version of LiteDB and open a new issue if your problem/question/suggestion still applies. Thanks!
